crates/alloy-dyn-contract/src/call.rs

```diff
@@ -0,0 +1,187 @@
+    pub const fn block(mut self, block: BlockId) -> Self {
+        self.block = Some(block);
+        self
+    }
+
+    /// Sets the [state override set](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set).
+    ///
+    /// # Note
+    ///
+    /// Not all client implementations will support this as a parameter to `eth_call`.
+    pub fn state(mut self, state: StateOverride) -> Self {
+        self.state = Some(state);
+        self
+    }
+
+    /// Returns the underlying transaction's ABI encoded data
+    pub fn calldata(&self) -> Option<Bytes> {
+        self.request.input.data.clone()
+    }
+}
+
+impl<P> CallBuilder<P>
+where
+    P: TempProvider,
+{
+    /// Returns the estimated gas cost for the underlying transaction to be executed
+    pub async fn estimate_gas(&self) -> Result<U256> {
+        self.provider.estimate_gas(self.request.clone(), self.block).await.map_err(Into::into)
+    }
+
+    /// Queries the blockchain via an `eth_call` for the provided transaction.
+    ///
+    /// If executed on a non-state mutating smart contract function (i.e. `view`, `pure`)
+    /// then it will return the raw data from the chain.
+    ///
+    /// If executed on a mutating smart contract function, it will do a "dry run" of the call
+    /// and return the return type of the transaction without mutating the state
+    ///
+    /// # Note
+    ///
+    /// This function _does not_ send a transaction from your account
+    pub async fn call(&self) -> Result<Vec<DynSolValue>> {
+        let bytes = if let Some(state) = &self.state {
+            self.provider
+                .call_with_overrides(self.request.clone(), self.block, state.clone())
+                .await
+                .map_err(Error::from)?
+        } else {
+            self.provider.call(self.request.clone(), self.block).await.map_err(Error::from)?
+        };
+
+        // decode output
+        let data = self.function.abi_decode_output(bytes.as_ref(), true)?;
+
+        Ok(data)
+    }
+
+    /// Signs and broadcasts the provided transaction
+    pub async fn send(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+impl<P> Clone for CallBuilder<P>
+where
+    P: Clone,
+{
+    fn clone(&self) -> Self {
+        CallBuilder {
+            request: self.request.clone(),
+            function: self.function.clone(),
+            block: self.block,
+            state: self.state.clone(),
+            provider: self.provider.clone(),
+        }
+    }
+}
+
+/// [`CallBuilder`] can be turned into a [`Future`] automatically with `.await`.
+///
+/// Defaults to calling [`CallBuilder::call`].
+impl<P> IntoFuture for CallBuilder<P>
+where
+    Self: 'static,
+    P: TempProvider,
+{
+    type Output = Result<Vec<DynSolValue>>;
+
+    #[cfg(target_arch = "wasm32")]
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output>>>;
+
+    #[cfg(not(target_arch = "wasm32"))]
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output> + Send>>;
```

Comment:
Could `impl Future` work here?

Reply:
that feature is still unstable unfortunately